local/run.sh

```diff
@@ -49,6 +49,39 @@ mc cp "/src/local/channel-rust-${channel}.toml" "local/static/dist/channel-rust-
 echo "==> detecting the last rustc commit on branch ${branch}"
 commit="$(git ls-remote "${RUSTC_REPO}" | grep "refs/heads/${branch}" | awk '{print($1)}')"
 
+# While the nightly and beta channels have the channel name as the "release" in
+# the archive names, the stable channel uses the actual Rust and Cargo version
+# numbers. This hacky piece of code detects them.
+if [[ "${channel}" = "stable" ]]; then
+    raw_url="https://raw.githubusercontent.com/rust-lang/rust/${commit}"
+
+    echo "==> loading rust version from src/version"
+    rust_release="$(curl --fail "${raw_url}/src/version" 2>/dev/null || true)"
+
+    # Legacy location of the Rust version. Outdated since 1.48.0.
+    if [[ "${rust_release}" = "" ]]; then
+        echo "==> loading rust version from src/bootstrap/channel.rs"
+        raw_release=$(curl --fail "${raw_url}/src/bootstrap/channel.rs" 2>&1 || true)
+        rust_release="$(echo "${raw_release}" | grep "CFG_RELEASE_NUM:" | sed -r 's/pub const CFG_RELEASE_NUM: &str = "([^"]+)";/\1/')"
+
+        if [[ "${rust_release}" = "" ]]; then
+            echo "ERR failed to get the version number"
+            exit 1
+        fi
+    fi
+
+    # Do our best to guess the cargo version
+    # This is kinda of a mess, yeah, and it will surely break. Sorry to whoever
+    # will have to fix this.
+    cargo_release="$(echo "${rust_release}" | awk '{split($0,a,".");a[2]+=1; print "0." a[2] "." a[3]}')"
```

Comment:
Seems like the more reliable thing to do is the same thing we did with channel.rs above but with src/tools/cargo/Cargo.toml, unless we haven't yet checked out submodules?

But I think we should really stop uploading cargo tarballs as 0.47.0, and instead just use the Rust version for them too. Cargo's "library" version really isn't that important.

Reply:
> Seems like the more reliable thing to do is the same thing we did with channel.rs above but with src/tools/cargo/Cargo.toml, unless we haven't yet checked out submodules?

That's not so easy, as we have not checked out submodules (we don't even have a rustc clone at this stage). For `src/version` and `src/bootstrap/channel.rs` I'm querying raw.githubusercontent.com, but those endpoints don't follow submodules. I tried looking into ways of working around that, but I wasn't successful.

> But I think we should really stop uploading cargo tarballs as 0.47.0, and instead just use the Rust version for them too. Cargo's "library" version really isn't that important.

Yeah, and unfortunately it's not just Cargo who's doing things like that. Still, we need those hacks at least until a PR changing this reaches stable.